### PR TITLE
Connect the UI with the pipeline for gathering tracepoint call stacks

### PR DESCRIPTION
**Pull request by Mahmood-Darwish**
This PR allows sending the info from the new options under the flag `tracepoint_callstack_collection` to TracerImpl. 

Bug: http://b/242841346

Screenshot: http://screenshot/C3A89V8AduXEXdn.png

### CODE REVIEW

File: src/LinuxTracing/TracerImpl.cpp [hunk 1]
```diff
@@ -75,7 +75,15 @@ TracerImpl::TracerImpl(
...
-
+  thread_state_change_callstack_collection_ =
+      capture_options.thread_state_change_call_stack_collection();
+  thread_state_change_callstack_method_ =
+      capture_options.thread_state_change_call_stack_unwinding_method();
+
+  ORBIT_LOG("%d %d", capture_options.thread_state_change_call_stack_collection(),
+            capture_options.thread_state_change_call_stack_unwinding_method());
+  ORBIT_LOG("%d %d", thread_state_change_callstack_collection_,
+            thread_state_change_callstack_method_);
```
[the remark this answers is not in the document]
Comment (Mahmood-Darwish) [reply src/LinuxTracing/TracerImpl.cpp]: This should already be removed.

File: src/LinuxTracing/TracerImpl.cpp [hunk 2]
```diff
@@ -492,7 +504,24 @@ static bool OpenFileDescriptorsAndRingBuffersForAllTracepoints(
...
+            case CaptureOptions::kDwarf:
+            default:
+              tracepoint_fd = tracepoint_with_stack_event_open(tracepoint_category, tracepoint_name,
+                                                               -1, cpu, stack_dump_size);
+          }
+          break;
+        case CaptureOptions::kNoThreadStateChangeCallStackCollection:
+        default:
+          tracepoint_fd = tracepoint_event_open(tracepoint_category, tracepoint_name, -1, cpu);
+      }
```
[the remark this answers is not in the document]
Comment (Mahmood-Darwish) [reply src/LinuxTracing/TracerImpl.cpp]: Done.

File: src/LinuxTracing/TracerImpl.cpp [hunk 3]
```diff
@@ -566,15 +592,51 @@ void TracerImpl::InitSwitchesStatesNamesVisitor() {
...
+          if (trace_thread_state_ || trace_context_switches_) {
+            tracepoints_to_open.emplace_back("sched", "sched_switch",
+                                             &sched_switch_with_callchain_ids_);
+          }
+          if (trace_thread_state_) {
+            // We also need task:task_newtask, but this is already opened by
+            // OpenThreadNameTracepoints.
+            tracepoints_to_open.emplace_back("sched", "sched_wakeup",
+                                             &sched_wakeup_with_callchain_ids_);
+          }
```
[the remark this answers is not in the document]
Comment (Mahmood-Darwish) [reply src/LinuxTracing/TracerImpl.cpp]: Done.

File: src/LinuxTracing/TracerImpl.cpp [hunk 4]
```diff
@@ -1311,21 +1386,22 @@ uint64_t TracerImpl::ProcessSampleEventAndReturnTimestamp(const perf_event_heade
     SchedWakeupPerfEvent event = ConsumeSchedWakeupPerfEvent(ring_buffer, header);
     DeferEvent(event);
 
+  } else if (is_sched_switch_with_callchain) {
+  } else if (is_sched_wakeup_with_callchain) {
+  } else if (is_sched_switch_with_stack) {
+  } else if (is_sched_wakeup_with_stack) {
```
[the remark this answers is not in the document]
Comment (Mahmood-Darwish) [reply src/LinuxTracing/TracerImpl.cpp]: Done, but I don't think I know the correct format of how I should be putting my name in the TODO statement.

File: src/LinuxTracing/TracerImpl.cpp [hunk 5]
```diff
@@ -1311,21 +1386,22 @@ uint64_t TracerImpl::ProcessSampleEventAndReturnTimestamp(const perf_event_heade
...
   } else if (is_amdgpu_sched_run_job_event) {
     AmdgpuSchedRunJobPerfEvent event = ConsumeAmdgpuSchedRunJobPerfEvent(ring_buffer, header);
     DeferEvent(std::move(event));
     ++stats_.gpu_events_count;
-
   } else if (is_dma_fence_signaled_event) {
     DmaFenceSignaledPerfEvent event = ConsumeDmaFenceSignaledPerfEvent(ring_buffer, header);
     DeferEvent(std::move(event));
     ++stats_.gpu_events_count;
-
```
[the remark this answers is not in the document]
Comment (Mahmood-Darwish) [reply src/LinuxTracing/TracerImpl.cpp]: Weird, I don't rememeber doing this, I've fixed it.

**Review by beckerhe — APPROVED** [review]
LGTM

File: src/CaptureClient/include/CaptureClient/ClientCaptureOptions.h [hunk 6]
```diff
@@ -29,6 +29,14 @@ struct ClientCaptureOptions {
   orbit_grpc_protos::CaptureOptions::UnwindingMethod unwinding_method =
       orbit_grpc_protos::CaptureOptions::UnwindingMethod::CaptureOptions_UnwindingMethod_kUndefined;
 
+  orbit_grpc_protos::CaptureOptions::ThreadStateChangeCallStackCollection
+      thread_state_change_callstack_collection =
+          orbit_grpc_protos::CaptureOptions::kThreadStateChangeCallStackCollectionUnspecified;
+
+  orbit_grpc_protos::CaptureOptions::UnwindingMethod
```
Comment (florian-kuebler) [inline src/CaptureClient/include/CaptureClient/ClientCaptureOptions.h:36]: I don't think we should offer a different unwinding method setting other than the "overall" one that is used for sampling.
It makes thinks easier for the user and also simplifies the code.

File: src/LinuxTracing/TracerImpl.cpp [hunk 7]
```diff
@@ -566,15 +587,34 @@ void TracerImpl::InitSwitchesStatesNamesVisitor() {
...
   std::vector<TracepointToOpen> tracepoints_to_open;
+  absl::flat_hash_set<uint64_t>* current_sched_switch_ids = &sched_switch_ids_;
+  absl::flat_hash_set<uint64_t>* current_sched_wakeup_ids = &sched_wakeup_ids_;
+  if (thread_state_change_callstack_collection ==
+          CaptureOptions::kThreadStateChangeCallStackCollection &&
+      thread_state_change_callstack_method_ == CaptureOptions::kDwarf) {
+    current_sched_switch_ids = &sched_switch_with_stack_ids_;
+    current_sched_wakeup_ids = &sched_wakeup_with_stack_ids_;
+  }
+  if (thread_state_change_callstack_collection ==
```
Comment (florian-kuebler) [inline src/LinuxTracing/TracerImpl.cpp:605]:
```suggestion
 else if (thread_state_change_callstack_collection ==
```

Comment (florian-kuebler) [reply src/LinuxTracing/TracerImpl.cpp:605]: also consider nesting the ifs and have an "if (thread_state_.... ==  ...CallstackCollection) then" surrounding.

File: src/LinuxTracing/TracerImpl.cpp [hunk 8]
```diff
@@ -613,7 +654,8 @@ bool TracerImpl::OpenGpuTracepoints(const std::vector<int32_t>& cpus) {
        {"amdgpu", "amdgpu_sched_run_job", &amdgpu_sched_run_job_ids_},
        {"dma_fence", "dma_fence_signaled", &dma_fence_signaled_ids_}},
       cpus, &tracing_fds_, GPU_TRACING_RING_BUFFER_SIZE_KB, &gpu_tracepoint_ring_buffer_fds_per_cpu,
-      &ring_buffers_);
+      &ring_buffers_, CaptureOptions::kNoThreadStateChangeCallStackCollection,
```
Comment (florian-kuebler) [inline src/LinuxTracing/TracerImpl.cpp:657]: Consider giving those parameters a default value, such that you don't need to pass them if not interested.

File: src/CaptureClient/include/CaptureClient/ClientCaptureOptions.h (hunk 6 above)
Comment (Mahmood-Darwish) [reply src/CaptureClient/include/CaptureClient/ClientCaptureOptions.h:36]: What do you think @beckerhe ?

Comment (beckerhe) [reply src/CaptureClient/include/CaptureClient/ClientCaptureOptions.h:36]: I agree. I also thought we concluded this will be only for testing purposes and is going to be removed later.

File: src/LinuxTracing/TracerImpl.cpp (hunk 7 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/TracerImpl.cpp]: Done.

File: src/GrpcProtos/capture.proto [hunk 9]
```diff
@@ -600,6 +599,8 @@ message ThreadStateSlice {
   // Otherwise, they should be 0 to minimize bandwidth usage.
   uint32 wakeup_tid = 7;
   uint32 wakeup_pid = 8;
+
+  uint32 CallStackId = 10;
```
Comment (beckerhe) [inline src/GrpcProtos/capture.proto:603]: This seems unrelated, right?

File: src/OrbitQt/CaptureOptionsDialog.cpp [hunk 10]
```diff
@@ -57,19 +57,8 @@ CaptureOptionsDialog::CaptureOptionsDialog(QWidget* parent)
   QObject::connect(ui_->threadStateCheckBox, qOverload<bool>(&QCheckBox::toggled), this,
                    [this](bool checked) {
```
Comment (beckerhe) [inline src/OrbitQt/CaptureOptionsDialog.cpp:58]:
```cpp
QObject::connect(ui_->threadStateCheckBox, qOverload<bool>(&QCheckBox::toggled), ui_->threadStateChangeCallstackCollectionCheckBox, &QCheckBox::setEnabled);
```

(No need for a lambda anymore.)

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

File: src/LinuxTracing/TracerImpl.cpp [hunk 11]
```diff
@@ -482,7 +483,12 @@ static bool OpenFileDescriptorsAndRingBuffersForAllTracepoints(
     const std::vector<TracepointToOpen>& tracepoints_to_open, const std::vector<int32_t>& cpus,
     std::vector<int>* tracing_fds, uint64_t ring_buffer_size_kb,
     absl::flat_hash_map<int32_t, int>* tracepoint_ring_buffer_fds_per_cpu_for_redirection,
-    std::vector<PerfEventRingBuffer>* ring_buffers) {
+    std::vector<PerfEventRingBuffer>* ring_buffers, uint32_t stack_dump_size,
+    const CaptureOptions::ThreadStateChangeCallStackCollection
+        thread_state_change_callstack_collection =
+            CaptureOptions::kNoThreadStateChangeCallStackCollection,
+    const CaptureOptions::UnwindingMethod thread_state_change_callstack_unwinding_method =
```
Comment (florian-kuebler) [inline src/LinuxTracing/TracerImpl.cpp:490]:
```suggestion
    const CaptureOptions::UnwindingMethod unwinding_method =
```

**Review by florian-kuebler — APPROVED** [review]
LGTM. Thanks!
